stdenv/linux: document some tips in debugging stdenv bootstrap tower

trofi opened the pull request:
Just a few comments added:

- added a few one-liners to explore which tools are rebuilt at each stdenv iteration during bootstrap
- explicitly listed available toolchains and their sources for on each
  bootstrap step: glibc, binutils, gcc, coreutils.

Should help slightly in explaining and untangling https://github.com/NixOS/nixpkgs/issues/208412

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

trofi committed:
stdenv/linux: document some tips in debugging stdenv bootstrap tower

Just a few comments added:

- added a few one-liners to explore which tools are rebuilt at each
  stdenv iteration during bootstrap
- explicitly listed available toolchains and their sources for on each
  bootstrap step: glibc, binutils, gcc, coreutils.
- added mention of static libraries linked into gcc

Co-authored-by: Adam Joseph <54836058+amjoseph-nixpkgs@users.noreply.github.com>

tpwrules commented:
There are some other utilities present in the final stdenv like xz and things rebuilt multiple times during the bootstrap like perl. Should those be addressed here?

trofi commented:
> There are some other utilities present in the final stdenv like xz and things rebuilt multiple times during the bootstrap like perl. Should those be addressed here?

Good question. Some of the secondary dependencies are volatile in nature: they come and go with patches and release updates.   It's quite a big list (dumped it below).

Some dependencies are very important, like `libidn2->glibc->libidn2` loop. Or a hack to clobber `ld`'s `INTERP` right after `glibc` upgrade. Some feel less relevant, like `xz`, `bzip` (and maybe `perl`?). Some are somewhat important, like static linkage of `libmpc` into final `gcc`, but not very special WRT which stage builds it.

We can pick more interesting cases and document them as well.

So far I mainly followed toolchain-like packages: those that produce code (gcc, binutils), or those that provide code for embedding (glibc).

The whole bootstrap list (needs some scrolling to the right to see the list):

```console
  * s0 -> s1+
    $ nix-store --query --graph $(nix-instantiate -A stdenv) | grep -P -v '[.]sh|[.]patch|bash|[.]tar' | grep -P '.*bootstrap-stage0-stdenv.*->.*'
    "sss8m2sbnqbmshm60n2qiaglnv5zrjqp-bootstrap-stage0-stdenv-linux.drv" -> "gxdaidcbnaamgldrsn3141vbm1irvrm3-bootstrap-stage2-gcc-wrapper-.drv" [color = "red"];
    "sss8m2sbnqbmshm60n2qiaglnv5zrjqp-bootstrap-stage0-stdenv-linux.drv" -> "iax324ji7951dqbj0fzwip8ds4vac324-bootstrap-stage0-glibc-bootstrap.drv" [color = "magenta"];
    "sss8m2sbnqbmshm60n2qiaglnv5zrjqp-bootstrap-stage0-stdenv-linux.drv" -> "kfp58s06f3vk1h4nnbdfpzpblf46ifk8-bootstrap-stage1-gcc-wrapper-.drv" [color = "blue"];
    "sss8m2sbnqbmshm60n2qiaglnv5zrjqp-bootstrap-stage0-stdenv-linux.drv" -> "0pid46kmxphfjmx7wlr9fdl7fllw29pa-bootstrap-stage0-binutils-wrapper-.drv" [color = "red"];
    "sss8m2sbnqbmshm60n2qiaglnv5zrjqp-bootstrap-stage0-stdenv-linux.drv" -> "ngdp6x4cfpcrwzn4gahkdxcy1hg0spk5-bootstrap-stage3-gcc-wrapper-.drv" [color = "black"];
    "sss8m2sbnqbmshm60n2qiaglnv5zrjqp-bootstrap-stage0-stdenv-linux.drv" -> "nkpkl0050sm25fqybik5qyc3z3fhbs3k-bootstrap-stage4-gcc-wrapper-13.0.0.drv" [color = "burlywood"];

  * s1 -> s2+
    $ nix-store --query --graph $(nix-instantiate -A stdenv) | grep -P -v '[.]sh|[.]patch|bash|[.]tar' | grep -P '.*bootstrap-stage1-stdenv.*->.*'
    "rpyd9fyc5idrihbik72z760zl7748d2h-bootstrap-stage1-stdenv-linux.drv" -> "1j9wzpdamixsi1j1amlrbycsma5kwa0n-gnum4-1.4.19.drv" [color = "red"];
    "rpyd9fyc5idrihbik72z760zl7748d2h-bootstrap-stage1-stdenv-linux.drv" -> "3ha7dr05jcib6irhv9sqm6j1cbix2x9s-bison-3.8.2.drv" [color = "black"];
    "iw8m0ajv0mk6rgdmvy1lwdnxylxlfsgh-bootstrap-stage1-stdenv-linux.drv" -> "3hwhcsn5azmzlxj2v1b5s0gj55kz0svm-binutils-wrapper-2.39.drv" [color = "black"];
    "rpyd9fyc5idrihbik72z760zl7748d2h-bootstrap-stage1-stdenv-linux.drv" -> "886hzg1wmj6rcwb7ry6xzc5smbpc4psd-expand-response-params.drv" [color = "red"];
    "rpyd9fyc5idrihbik72z760zl7748d2h-bootstrap-stage1-stdenv-linux.drv" -> "0pg4y7baxc39szjr5d6ka8g028z5f4w1-binutils-2.39.drv" [color = "red"];
    "rpyd9fyc5idrihbik72z760zl7748d2h-bootstrap-stage1-stdenv-linux.drv" -> "1d6ddz6k73d1yiifvyha7d5xi1mr0ada-gettext-0.21.drv" [color = "red"];
    "rpyd9fyc5idrihbik72z760zl7748d2h-bootstrap-stage1-stdenv-linux.drv" -> "ak0ybs7w05hjffziki3qngjv8020r6kz-texinfo-6.8.drv" [color = "green"];
    "rpyd9fyc5idrihbik72z760zl7748d2h-bootstrap-stage1-stdenv-linux.drv" -> "f7hs58r69qlp3bv03578yqx01nfb2lxy-xz-5.4.0.drv" [color = "red"];
    "iw8m0ajv0mk6rgdmvy1lwdnxylxlfsgh-bootstrap-stage1-stdenv-linux.drv" -> "igzsn1dzng4bfb7l4pqzq3a3l8mwqiqi-binutils-wrapper-2.39.drv" [color = "green"];
    "rpyd9fyc5idrihbik72z760zl7748d2h-bootstrap-stage1-stdenv-linux.drv" -> "qigxgh2h1xm90xgvqcjqcd9v2q6hkxmh-zlib-1.2.13.drv" [color = "burlywood"];
    "rpyd9fyc5idrihbik72z760zl7748d2h-bootstrap-stage1-stdenv-linux.drv" -> "shi17zrfdag2cw437lz4lfbfby7y0gn3-perl-5.36.0.drv" [color = "burlywood"];

 * s2-> s3+
   $ nix-store --query --graph $(nix-instantiate -A stdenv) | grep -P -v '[.]sh|[.]patch|bash|[.]tar' | grep -P '.*bootstrap-stage2-stdenv.*->.*'
   "89kaz6k451ck2vh8ga2valgw4wq130j4-bootstrap-stage2-stdenv-linux.drv" -> "17lzy22lvw9w8qvp18dh1zf8j8b1kim1-linux-headers-6.1.drv" [color = "black"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "57271y6vc1353w92fxkq58wj7h49gfmz-libidn2-2.3.2.drv" [color = "green"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "5kivra808dsk9wmyvqjzv91gq0x9gcsi-libunistring-1.0.drv" [color = "red"];
   "89kaz6k451ck2vh8ga2valgw4wq130j4-bootstrap-stage2-stdenv-linux.drv" -> "8id9zhcpijc4l2iy59bq376yks5z8k98-bootstrap-stage0-glibc-iconv-bootstrap.drv" [color = "blue"];
   "89kaz6k451ck2vh8ga2valgw4wq130j4-bootstrap-stage2-stdenv-linux.drv" -> "a5aab34ih7hjz7dkj072p34pp5p4k657-binutils-2.39.drv" [color = "burlywood"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "mbr3wk8xdhibfb0gqljnls92b69skypv-expand-response-params.drv" [color = "burlywood"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "savk0l1zv1wiz32l30rwgfymbm0klwsx-patchelf-0.15.0.drv" [color = "black"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "vijzw99ib8lx2zzkd42agb0lws8f60ki-glibc-2.35-224.drv" [color = "burlywood"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "vw7jj8i48zjld8g8xxrvng60cgd7gndf-python3-minimal-3.10.8.drv" [color = "magenta"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "1c1j06fjz38afd5vk7dlc53q0bikv7gw-autoconf-archive-2022.09.03.drv" [color = "burlywood"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "34nfl1ijs461rcgyskn5jabjsn7q9y5a-xz-5.4.0.drv" [color = "green"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "56jffxbkyviwjjdah1wd8q6cqwgk38sf-bzip2-1.0.8.drv" [color = "red"];
   "89kaz6k451ck2vh8ga2valgw4wq130j4-bootstrap-stage2-stdenv-linux.drv" -> "80ywjgkdkswcsxfnd21h0v4zwwq52a0r-hook.drv" [color = "green"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "0cvwyqrfl8hf1z1dch7p1869lgbhz95m-libtool-2.4.7.drv" [color = "burlywood"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "1y62inb9g43nk1qbhg8w7lbccpn7kfm5-autoconf-2.71.drv" [color = "blue"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "7cfgkhd876ipbqs4xdj93mvfpc2hrwlq-automake-1.16.5.drv" [color = "black"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "cfkkc6m38prsdbyzvyj61vay2x6fhl8g-zlib-1.2.13.drv" [color = "magenta"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "cxmdps9l7qxkaq999risiynlb7m5f8fx-gettext-0.21.drv" [color = "black"];
   "89kaz6k451ck2vh8ga2valgw4wq130j4-bootstrap-stage2-stdenv-linux.drv" -> "d588xih5x1aj0zps766ixbrcykk67i04-pkg-config-wrapper-0.29.2.drv" [color = "red"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "gq5vazp93cz5rb9fz2dbk523axpf4wy0-expat-2.5.0.drv" [color = "red"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "gvqv2bk6b0qxsji9vxdhm5w3f7vsc7ra-pkg-config-0.29.2.drv" [color = "black"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "hvq0c8dp4yy7rv7b2rr5mgm0k4wqpjq2-libxcrypt-4.4.33.drv" [color = "magenta"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "md5yc9cvr88hn6b8k68j8mdqq9jv846m-texinfo-6.8.drv" [color = "black"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "qpihwmpl07jy7kx6dsh0wcv7ryg5j9kc-help2man-1.49.2.drv" [color = "blue"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "bnj413zi1gm425z9fjkk9z1mxx1k8vrr-perl5.36.0-gettext-1.07.drv" [color = "green"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "c8h7vlknl1gfh0z21cc32z26fba6ryj7-perl-5.36.0.drv" [color = "red"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "x1125aihymyqrs5d9wdmcryf2amd1pb3-file-5.43.drv" [color = "black"];
   "89kaz6k451ck2vh8ga2valgw4wq130j4-bootstrap-stage2-stdenv-linux.drv" -> "yvw7zhr0kq3qaa1c9wa0s2hpfgmwqnlm-nuke-references.drv" [color = "red"];
   "kd72va0i3ay01hcszk4rrp4agpqjzgn2-bootstrap-stage2-stdenv-linux.drv" -> "z250qq4j5pbjxqa49qzf6hlsa5c07gk9-libffi-3.4.4.drv" [color = "black"];

  * s3 -> s4+
    $ nix-store --query --graph $(nix-instantiate -A stdenv) | grep -P -v '[.]sh|[.]patch|bash|[.]tar' | grep -P '.*bootstrap-stage3-stdenv.*->.*'
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "2dx7k86vw2701cidbwlf5ax91387n5lp-gettext-0.21.drv" [color = "burlywood"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "c8jfkzv5z1wyri93f3zxi9bsld5f4bqv-texinfo-6.8.drv" [color = "black"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "cv31bwgvz1ms8g481y0hk37lrn6y6sx0-zlib-1.2.13.drv" [color = "green"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "d47js1wik2lf9j9x0rn59vnjfpa6kir3-gcc-13.0.0.drv" [color = "red"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "8ngy9ncv45y6lpjsxkbbdhd2w1i667vb-mpfr-stage3-4.1.1.drv" [color = "burlywood"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "didv4h48bp73j5w2swd8v9lbs58yjzqn-libmpc-stage3-1.2.1.drv" [color = "magenta"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "bkwxrqy1xw1plrjgbwr3686frr4qfdwa-mpfr-4.1.1.drv" [color = "blue"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "fhwr5c6f6kd23045ib9fi0qr0m5pa1xh-expand-response-params.drv" [color = "green"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "iqh4dg54hkpzc4cc3wwmsa3frqacdikk-xz-5.4.0.drv" [color = "green"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "msmp6hkkrmipivaibyn7r8llhq81zr0x-flex-2.6.4.drv" [color = "blue"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "hxwl3f6kbsm2d80gn8lnvl23gzsm51xr-help2man-1.49.2.drv" [color = "red"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "9y5490350svrxysfzz1x8cirpv80q030-perl5.36.0-gettext-1.07.drv" [color = "black"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "igh4cwfbwrcd43sq4a64rkqzimzzvhpj-perl-5.36.0.drv" [color = "red"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "p31hzp67h1ir41j3avxxj1y31v8c7lj1-gmp-with-cxx-6.2.1.drv" [color = "burlywood"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "s46jnhcswf71wds6qi0k5fb7rwsjvxm5-which-2.21.drv" [color = "green"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "vp3r1q7h50bfbdfs5sca1v7v4bq2z8h5-libxcrypt-4.4.33.drv" [color = "green"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "wdy9ci6q8p8w03yw4jp9za4s2qfkgyna-isl-stage3-0.20.drv" [color = "burlywood"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "wwavja7952gn82c6vd60bp861jfmnqvk-gmp-with-cxx-stage3-6.2.1.drv" [color = "black"];
    "9ics3hn931bfxmll6pga3ixb9vpir3dh-bootstrap-stage3-stdenv-linux.drv" -> "x2zfsc8gb1x8s4qcgahlgmrx6rfd74qb-hook.drv" [color = "green"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "512x62mw6b2b94j24dqmqcrqvp003dgb-libtool-2.4.7.drv" [color = "blue"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "51m3rnl383w7x0k1ahvdzzj2cwrfmg9p-autoconf-2.71.drv" [color = "blue"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "7aq50zyqmhav7z712w95693vcsw28q6v-automake-1.16.5.drv" [color = "burlywood"];
    "chga8a1d6pcxk4i1yv40n4bzmr0vmjzv-bootstrap-stage3-stdenv-linux.drv" -> "gv67lrmir7r5k5rq2lis8c2rhf1lzl4l-file-5.43.drv" [color = "green"];

  * s4 -> final
    $ nix-store --query --graph $(nix-instantiate -A stdenv) | grep -P -v '[.]sh|[.]patch|bash|[.]tar' | grep -P '.*bootstrap-stage4-stdenv.*->.*'
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "2kkpn9l97lfryd8alvjy3p923y2aw3cn-attr-2.5.1.drv" [color = "burlywood"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "39dw118hq02bx4x4v3ay2p6jc287qj2j-gmp-with-cxx-stage4-6.2.1.drv" [color = "green"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "6f7x7grg4ync7w1qrgmq1bmy2239g5s9-diffutils-3.8.drv" [color = "blue"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "6qd137375hd066w6lwwj8s297mkzn8bx-bzip2-1.0.8.drv" [color = "green"];
    "f77mcaqvynmin9f77z06sziirlxfk7p5-bootstrap-stage4-stdenv-linux.drv" -> "5gqsa7vc2h9h1h3jgpw4ybxcs8f0l8ib-hook.drv" [color = "magenta"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "35xzppkmbbijs29kbshimmi7l2abckv7-automake-1.16.5.drv" [color = "burlywood"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "3y275vwm6s5wab603jb8s3ddgscfg59c-autoconf-2.71.drv" [color = "red"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "6pj2g2s3dcczb70ql0aypapb3f3lqasq-libtool-2.4.7.drv" [color = "burlywood"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "9bh10gs9zhhkjhvijpdxyapm7xh0ks21-patchelf-0.15.0.drv" [color = "green"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "b66qnf4bfy57a80g6ijpv5bvs17w9j2n-gnused-4.9.drv" [color = "black"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "c4bsk8sw0im51jdmkbh2jzxa17wiyxba-gnugrep-3.7.drv" [color = "black"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "c5s9fm842njyilahi3fx9ff9slz4933q-gnutar-1.34.drv" [color = "red"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "d6096q9lzkpsavpv88f7xg762djz0mpr-help2man-1.49.2.drv" [color = "blue"];
    "f77mcaqvynmin9f77z06sziirlxfk7p5-bootstrap-stage4-stdenv-linux.drv" -> "dlclsqrd208ry7imchhk8rpd8rz2w01g-gcc-wrapper-13.0.0.drv" [color = "red"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "fxmg8l1l2rbw0l9b5w31n5n2qw7ysimi-xz-5.4.0.drv" [color = "black"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "g0mh37ih39wwyrrb8fj7qrq86pa6n68b-coreutils-9.1.drv" [color = "blue"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "ghv09314nh7khp77igjcy02162swqlyf-perl5.36.0-gettext-1.07.drv" [color = "burlywood"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "hmiky549ypqxxmvg9r0arax2v13idvd9-gawk-5.2.1.drv" [color = "burlywood"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "i2lnvviydhlxmz6g1s65cg1bycxyw4sy-perl-5.36.0.drv" [color = "blue"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "irhgkpxibylrxbbnf5nfn6r83fi3yk3i-patch-2.7.6.drv" [color = "burlywood"];
    "f77mcaqvynmin9f77z06sziirlxfk7p5-bootstrap-stage4-stdenv-linux.drv" -> "iw78x2ijj2cvnw1w21sn0qnhpgj3g8b8-binutils-wrapper-2.39.drv" [color = "burlywood"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "k3l1yrp4jq5w1gv1dbkkrz6r1vab1bww-gzip-1.12.drv" [color = "magenta"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "kpxx5lgj0qjviav2bfzaml8iqi43yk7w-gnumake-4.4.drv" [color = "blue"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "m4y5pfnvmaqlxjb8gdn4jhg3hbp8vlbb-file-5.43.drv" [color = "green"];
    "f77mcaqvynmin9f77z06sziirlxfk7p5-bootstrap-stage4-stdenv-linux.drv" -> "r71xhg2hvsnqk6lj4lpzbga6321w0xsd-setup-hook.drv" [color = "green"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "r9qx3z26l80x90z1wmz7rbfxx1bchwcn-ed-1.18.drv" [color = "black"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "pcbm3np7rhzsgdhwcwakxlwka3lp0v5a-lzip-1.23.drv" [color = "burlywood"];
    "f77mcaqvynmin9f77z06sziirlxfk7p5-bootstrap-stage4-stdenv-linux.drv" -> "srdmshnbmgi13nsgmaw3m8a79sinadqr-glibc-iconv-2.35.drv" [color = "green"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "vlwl4vrjqqi03xlnxizfkkxm5hffif0g-acl-2.3.1.drv" [color = "blue"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "x3iclzi0q0dqq66ah269r082axnfavwy-binutils-2.39.drv" [color = "red"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "xh4ngz24lbzwsr8yqjzbdzp6myikg0bq-findutils-4.9.0.drv" [color = "burlywood"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "xr735lr697xkk59hinnkb05ir528qcav-libxcrypt-4.4.33.drv" [color = "magenta"];
    "g85bilaynpw3dfg3bf0yd7lpwa60kmri-bootstrap-stage4-stdenv-linux.drv" -> "xsqc2qi99dmvnaiks6sffznqysjdrhvb-pcre-8.45.drv" [color = "green"];
```

tpwrules commented:
Looking at this again, I'd say you do actually address the other utilities implicitly in the new section on bootstrap goals. The goals sound good to me. Maybe add some justification like "These goals ensure the final stdenv has been built exclusively by packages available in the final stdenv so that compiler versions and environments etc. are consistent". That at least is how I understand the goals.

Is there any mechanical way to check that those those goals are met? I think the `allowedRequisites` and `disallowedRequisites` will ensure point 1 is not violated by accident (though we currently violate it deliberately using `cp`; checking hashes should catch that). I'm not sure if there's a way to do point 2 except by manual inspection, and I'm not immediately sure what to look for.

Can you put the command to generate that list in the comments too? That looks very helpful.

trofi commented:
> Looking at this again, I'd say you do actually address the other utilities implicitly in the new section on bootstrap goals. I just wanted to be sure that we knew where they should come from. The goals sound good to me. Maybe add some justification like "These goals ensure the final stdenv has been built exclusively by packages available in the final stdenv so that compiler versions and environments etc. are consistent". That at least is how I understand the goals.

I think I get what you say. I added the following blurb:

```
# These goals ensure that final packages and final stdenv are built
# exclusively using nixpkgs package definitions and don't depend
# on bootstrapTools (via direct references or via inclusion
# of copied code of generated code by bootstrapTools).
```

I think we are on the same page, but just in case I'll try to explain myself a bit.

I did not include your definition as is because (AFAIU) there is a fundamental problem of bootstrap process that has to be solved by pulling package definitions from non-final stdenv:

- most of `pkgs` depends on `stdenv` (by literally depending on things like `gcc` from `stdenv`)
- `stdenv` depends on `pkgs` (version of `gcc` from previous stage of `stdenv`)

To untie this infinitely unfolding cycle final `stdenv` stage overrides some of normal package definitions to refer to package definitions of previous `stdenv`. This make latest `stdenv` to use non-vanilla package definitions.

The final result is morally equivalent to your phrasing: "final stdenv has been built exclusively by packages available in the final stdenv". But mechanically we do it the other way around: we declare that previous (`stage4` instance) is good enough and untied enough from `bootstrapTools` to declare it final and copy some of it's reference into `final` stdenv as is. While in reality it's built from various `stdenv` iterations. We expect them to be the same if we add more rebuild iterations, but we can't easily ensure it's always true.

Thus the more vague wording around the goals.

I hope I did not make things worse with this brain dump.

> Is there any mechanical way to check that those those goals are met? I think the `allowedRequisites` and `disallowedRequisites` will ensure point 1 is not violated by accident (though we currently violate it deliberately using `cp`; checking hashes should catch that). I'm not sure if there's a way to do point 2 except by manual inspection, and I'm not immediately sure what to look for.

I don't think there is a good way to detect even our current `libgcc_s.so` hack :( It does not have a hash within that refers to a /nix/store path. The only hints are stale `RUNPATH` entries to nuked references and old compiler ID  in `.notes` sections. Both would be absent if we were a bit more eager removing these optional bits when we build `bootstrapTools` seed tarball.

`cp` is hard to detect reliably in general: for contentAddressed builds it's natural when produced binaries are identical to binaries produced on previous steps.

> Can you put the command to generate that list in the comments too? That looks very helpful.

Sounds good. Added.

tpwrules commented:
Rephrasing the blurb slightly and fixing the typo:

```
# These goals ensure that final packages and final stdenv are built
# exclusively using nixpkgs package definitions and don't depend
# on bootstrapTools (via direct references, inclusion
# of copied code, or code compiled directly by bootstrapTools).
```

I think this PR is good other than some other minor typos.

> I don't think there is a good way to detect even our current libgcc_s.so hack :( It does not have a hash within that refers to a /nix/store path. The only hints are stale RUNPATH entries to nuked references and old compiler ID in .notes sections. Both would be absent if we were a bit more eager removing these optional bits when we build bootstrapTools seed tarball.

I did not actually check, but I guess it is not a given that the file, especially since it's a binary, would make it through without being changed. That is unfortunate but oh well.

trofi commented:
> Rephrasing the blurb slightly and fixing the typo:
> 
> ```
> # These goals ensure that final packages and final stdenv are built
> # exclusively using nixpkgs package definitions and don't depend
> # on bootstrapTools (via direct references, inclusion
> # of copied code, or code compiled directly by bootstrapTools).
> ```

Applied this wording.

> I think this PR is good other than some other minor typos.

 Thank you!

amjoseph-nixpkgs commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -1,8 +1,55 @@
+# nixpkgs instances (stages) to gradually rebuilt bootstrapFiles
+# with nixpkgs equivalents of packages.
+#
+# Goals of the bootstrap process:
```
```suggestion
# Goals of the bootstrap process:
# 0. final stdenv must not reference any of the bootstrap files.
```
This is a separate goal from goal number 1 (on the next line), and stdenv already has achieved the first one.  Importantly, `allowedRequisites` provides an automated check that guarantees we won't accidentally lose this goal without noticing.

I fully support your more-ambitious goal (from #187225) of getting rid of the silly "copy libgcc_s hack".  If you manage to accomplish that, do you have any ideas on how we could add an automated check that would prevent people from copying things out of the bootstrap-files into stdenv?

amjoseph-nixpkgs commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -1,8 +1,55 @@
+# Goals of the bootstrap process:
+# 1. final stdenv must not contain any of the bootstrap files
+#    (the only current violation is libgcc_s.so in glibc).
+# 2. final stdenv must not contain any binaries generated directly
+#    by bootstrap compiler (the only currently known violation is
+#    all of glibc).
```
```suggestion
# 2. final stdenv must not contain any of the bootstrap files
#    (the only current violations are: libgcc_s.so in glibc,
#    the lib{mpfr,mpc,gmp,isl} which are statically linked
#    into the final gcc).
```

See `bootstrap-stage3`, where we use the bootstrap compiler is used to compile `.a` versions of those four libraries.  This drives me nuts, by the way.  I don't know why, but it bothers me a lot.  The code has been there since cfde88976ba4cddd01b1bb28b40afd12ea93a11d.

`nix-tree --derivation $(nix-instantiate -A gcc)` is helpful for untangling all of this.

trofi replied:
Good point on those static libs!

amjoseph-nixpkgs replied:
> Good point on those static libs!

I remembered why this bothers me so much: https://github.com/NixOS/nixpkgs/pull/170215/commits/518353b4b616989dd440cae20298502f245accc9

amjoseph-nixpkgs commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -1,8 +1,55 @@
+# - stdenv.__bootPackages.stdenv.__bootPackages.stdenv
+# - ... and so on.
+#
+# To explore build-time dependencies in graphincal form one can use
```
```suggestion
# To explore build-time dependencies in graphical form one can use
```

amjoseph-nixpkgs commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -1,8 +1,55 @@
+#       grep -P -v '[.]sh|[.]patch|bash|[.]tar' |
+#       grep -P '.*stdenv.*->.*glibc-2'
+#     "...-bootstrap-stage2-stdenv-linux.drv" -> "...-glibc-2.35-224.drv";
+
```
```suggestion
#
# For a TUI (rather than CLI) view, you can use:
#
#   nix-tree --derivation $(nix-instantiate -A stdenv)
```

trofi replied:
TIL about `nix-tree`!

amjoseph-nixpkgs commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -332,6 +393,14 @@ in
+  # - gcc: from nixpkgs, built by bootstrapFiles toolchain. Can assume
+  #        it has almost no code from bootstrapTools as gcc bootstraps
+  #        internally. The only exceptions are crt files from glibc
+  #        built by bootstraTools used to link executables and libraries.
```
```suggestion
  #        built by bootstraTools used to link executables and libraries,
  #        and the bootstrapTools-built, statically-linked
  #        lib{mpfr,mpc,gmp,isl}.a which are linked into the final gcc
  #        (see commit cfde88976ba4cddd01b1bb28b40afd12ea93a11d).
```

amjoseph-nixpkgs commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -388,6 +457,14 @@ in
+  # - gcc: from nixpkgs, built by bootstrapFiles toolchain. Can assume
+  #        it has almost no code from bootstrapTools as gcc bootstraps
+  #        internally. The only exceptions are crt files from glibc
+  #        built by bootstraTools used to link executables and libraries.
```
```suggestion
  #        built by bootstraTools used to link executables and libraries,
  #        and the bootstrapTools-built, statically-linked
  #        lib{mpfr,mpc,gmp,isl}.a which are linked into the final gcc
  #        (see commit cfde88976ba4cddd01b1bb28b40afd12ea93a11d).
```

amjoseph-nixpkgs reviewed (approved):
Hey, thanks for documenting all of this!

Except for `lib{mpfr,mpc,isl,gmp}.a` all of my review comments are just typos and minor stuff.  But please do take a look at the comments about how those are handled so we don't mislead people into thinking they are built by the final gcc['s stage1 bootstrap compiler] -- they aren't!

K900 commented:
Can we also get rid of the libmpfr hack if we add a second gcc rebuild? Seems like it, so it might actually be worth it.

amjoseph-nixpkgs commented:
> Can we also get rid of the libmpfr hack if we add a second gcc rebuild? Seems like it, so it might actually be worth it.

The ideal solution is to make gcc's stage1 and stage2 be separate nix derivations.  then we don't have to build gcc twice (or rather we don't have to build it *four times*).  I think @Ericson2314 had some thoughts on this topic.

trofi commented:
> Hey, thanks for documenting all of this!
> 
> Except for `lib{mpfr,mpc,isl,gmp}.a` all of my review comments are just typos and minor stuff. But please do take a look at the comments about how those are handled so we don't mislead people into thinking they are built by the final gcc['s stage1 bootstrap compiler] -- they aren't!

Thank you, Adam! I applied (and squashed) all your suggestions as is. Great point on gcc's static libraries! The are very important to cal out.

tpwrules commented:
Something got mangled with the goals, 1 and 2 are almost the same now. I think 2 should be the one about being compiled by the bootstrap compiler?

trofi commented:
> Something got mangled with the goals, 1 and 2 are almost the same now. I think 2 should be the one about being compiled by the bootstrap compiler?

Good catch! Re-numbered to start from `1.` and reworded last goal in terms of code generators:

```
...
# 2. final stdenv must not contain any of the bootstrap files
#    (the only current violation is libgcc_s.so in glibc).
# 3. final stdenv must not contain any of the files directly
#    generated by the bootstrap code generators (assembler, linker,
#    compiler).
```

Ericson2314 commented:
Thanks for doing this! I do want to make it saner per things like building GCC separately just as @amjoseph-nixpkgs said, but yes let that not be a reason not to document how things exist currently. I learned stuff reading this too!
